### DISCUSSION TITLE
How to invoke command if multiple previous commands execute successfully?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [PowerShell/PowerShell#21330] [Q&A] [answered]
I'm currently performing:

```.PS1
sudo zypper dup && `
flatpak update -y && `
flatpak uninstall --unused && `
sudo snap refresh && `
Stop-Computer
```

I'd like to solely invoke `Stop-Computer` after *all* of its predecessors have executed successfully as it does now, but I don't want any of its predecessors to be dependent upon *their* predecessor's successful invocation.

Specifically, `Stop-Computer` should solely be invoked if `sudo zypper dup`, `flatpak update -y`, `flatpak uninstall --unused` and `sudo snap refresh` have completed successfully. However, `flatpak update -y` should be executed regardless of whether `sudo zypper dup` did successfully, as `sudo snap refresh` should be invoked regardless of whether `flatpak uninstall --unused` completes.

I'm probably going the wrong way about doing this. Maybe I should just be using a function which is called before `Stop-Computer`, allowing me to read *its* `$?` instead?

**Comment by rhubarb-geek-nz** [comment]
You could use jobs to do this, each job will run but the final statement will only run if none had an error

```
#!/usr/bin/env pwsh

$ErrorActionPreference = "Stop"

trap
{
        throw $PSItem
}

$null = Start-Job -ScriptBlock {
        Start-Sleep 5
        Write-Output "sudo zypper dup"
}

$null = Start-Job -ScriptBlock {
        Start-Sleep 10
        Write-Output "flatpak update -y"
        Write-Output "flatpak uninstall --unused"
        # Write-Error "foo"
}

$null = Start-Job -ScriptBlock {
        Start-Sleep 15
        Write-Output "sudo snap refresh"
}

Get-Job | Wait-Job
Get-Job | Receive-Job

Write-Output "Stop-Computer"
```

Will produce

```
Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
--     ----            -------------   -----         -----------     --------             -------
1      Job1            BackgroundJob   Completed     True            localhost            …
3      Job3            BackgroundJob   Completed     True            localhost            …
5      Job5            BackgroundJob   Completed     True            localhost            …
sudo zypper dup
flatpak update -y
flatpak uninstall --unused
sudo snap refresh
Stop-Computer
```

If you uncomment the Write-Error you get

```
Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
--     ----            -------------   -----         -----------     --------             -------
1      Job1            BackgroundJob   Completed     True            localhost            …
3      Job3            BackgroundJob   Completed     True            localhost            …
5      Job5            BackgroundJob   Completed     True            localhost            …
sudo zypper dup
flatpak update -y
flatpak uninstall --unused
Exception:  .../waitjob.ps1:28
Line |
  28 |  Get-Job | Receive-Job
     |  ~~~~~~~~~~~~~~~~~~~~~
     | foo
```

**Comment by rhubarb-geek-nz** [comment, marked as answer]
Given you are running external programs and have to check the $LastExitCode to detect an error then sum up the $LastExitcCde, it it is zero then there were no errors

```
#!/usr/bin/env pwsh

$ErrorActionPreference = "Stop"

trap
{
        throw $PSItem
}

$SumLastExitCode = 0

Write-Output "sudo zypper dup"

/usr/bin/true
$SumLastExitCode += $LastExitCode

Write-Output "flatpak update -y"
Write-Output "flatpak uninstall --unused"

/usr/bin/true
$SumLastExitCode += $LastExitCode

Write-Output "sudo snap refresh"

/usr/bin/true
$SumLastExitCode += $LastExitCode

if ( $SumLastExitCode -eq 0)
{
        Write-Output "Stop-Computer"
}
```

Change any of the /usr/bin/true to /usr/bin/false to see the effect

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21330#discussioncomment-8754359

@rhubarb-geek-nz, that might be one of the most intuitive demonstrations of a programmatic concept I've ever had the pleasure of using. I'm seriously thankful. I've used it as https://gitlab.com/rokejulianlockhart/rokejulianlockhart/-/blob/61bbdd7d0ba1963c2723467b1850ed616ae7a5c2/%09%60%7Btitle:%20%22Entities%22,%09id:%20%22s1gvg2%22%7D%60%7B.JSON5%7D.dir/%09%60%7Bname:%20%7Bshort:%20%22Beedell.RokeJulianLockhart%22,%09long:%20%7Bthird:%20%22Beedell%22,%09first:%20%22Roke%22%7D%7D,%09id:%20%22s1gvgt%22%7D%60%7B.JSON5%7D.dir/%09%60%7Btitle:%20%22Technical%22,%09id:%20%22s20cxw%22%7D%60%7B.JSON5%7D.dir/%09%60%7Btitle:%20%22Command%20Reference%22,%09id:%20%22%22%7D%60%7B.JSON5%7D.txt.MD?plain=1#L76-108 demonstrates.
